File: crates/qks-plugin/bindings/swift/Sources/QKS/FFI.swift
/**
 * Low-level FFI bindings to QKS Rust library using C interop.
 *
 * DO NOT use these functions directly - use the high-level APIs instead.
 */

import Foundation

// MARK: - C Function Declarations

/// Opaque handle to QKS instance
public typealias QksHandle = OpaquePointer

/// QKS configuration (C struct)
struct QksConfigC {
    var numQubits: size_t
    var useGpu: Bool
    var enableConsciousness: Bool
    var enableMetacognition: Bool
}

/// IIT Φ computation result (C struct)
struct PhiResultC {
    var phi: Double
    var mipSize: size_t
    var numStates: size_t
    var computationTimeMs: Double
}

/// Metacognitive introspection report (C struct)
struct IntrospectionReportC {
    var numBeliefs: size_t
    var numGoals: size_t
    var numCapabilities: size_t
    var confidence: Double
    var timestampMs: UInt64
}

// MARK: - Core Functions

@_silgen_name("qks_create")
func qks_create(_ config: UnsafePointer<QksConfigC>?) -> QksHandle?

@_silgen_name("qks_destroy")
func qks_destroy(_ handle: QksHandle)

@_silgen_name("qks_version")
func qks_version() -> UnsafePointer<CChar>

// MARK: - Layer 6: Consciousness

@_silgen_name("qks_compute_phi")
func qks_compute_phi(
    _ handle: QksHandle,
    _ networkState: UnsafePointer<Double>,
    _ stateSize: size_t,
    _ result: UnsafeMutablePointer<PhiResultC>
) -> Int32

@_silgen_name("qks_broadcast")
func qks_broadcast(
    _ handle: QksHandle,
    _ content: UnsafeRawPointer,
    _ contentSize: size_t,
    _ priority: Double
) -> Int32

// MARK: - Layer 7: Metacognition

@_silgen_name("qks_introspect")
func qks_introspect(
    _ handle: QksHandle,
    _ report: UnsafeMutablePointer<IntrospectionReportC>
) -> Int32

@_silgen_name("qks_monitor_performance")
func qks_monitor_performance(
    _ handle: QksHandle,
    _ metrics: UnsafePointer<Double>,
    _ numMetrics: size_t
) -> Int32

// MARK: - Layer 8: Integration

@_silgen_name("qks_cognitive_cycle")
func qks_cognitive_cycle(
    _ handle: QksHandle,
    _ sensoryInput: UnsafeRawPointer,
    _ inputSize: size_t,
    _ outputBuffer: UnsafeMutableRawPointer,
    _ outputSize: size_t
) -> Int32

// MARK: - Error Handling

@_silgen_name("qks_get_last_error")
func qks_get_last_error() -> UnsafePointer<CChar>?

/// QKS error type
public enum QKSError: Error, CustomStringConvertible {
    case initializationFailed(String)
    case computationFailed(String)
    case invalidInput(String)
    case nativeError(String)

    public var description: String {
        switch self {
        case .initializationFailed(let msg):
            return "Initialization failed: \(msg)"
        case .computationFailed(let msg):
            return "Computation failed: \(msg)"
        case .invalidInput(let msg):
            return "Invalid input: \(msg)"
        case .nativeError(let msg):
            return "Native error: \(msg)"
        }
    }
}

/// Check FFI result code and throw error if failed
func checkError(_ result: Int32) throws {
    if result != 0 {
        var errorMessage = "QKS error code: \(result)"
        if let errorPtr = qks_get_last_error() {
            errorMessage = String(cString: errorPtr)
        }
        throw QKSError.nativeError(errorMessage)
    }
}

// MARK: - Library Loading

/// Load the QKS shared library
func loadQKSLibrary() -> Bool {
    // The library should be linked at compile time or loaded via dlopen
    // For now, assume it's available through proper linking
    return true
}

File: crates/qks-plugin/bindings/typescript/src/ffi.ts
/**
 * Low-level FFI bindings to QKS Rust library using Node-API.
 *
 * DO NOT use these functions directly - use the high-level APIs instead.
 *
 * @internal
 */

import { createRequire } from 'module';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

// Determine library path
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

let _native: any;

/**
 * Load the native QKS library.
 * Searches in order: release build, debug build, installed location.
 */
function loadNative(): any {
    if (_native) return _native;

    const searchPaths = [
        // Development paths (relative to bindings)
        join(__dirname, '../../../../target/release'),
        join(__dirname, '../../../../target/debug'),
        // Installed path (node_modules)
        join(__dirname, '../native'),
    ];

    let lastError: Error | null = null;

    for (const searchPath of searchPaths) {
        try {
            const require = createRequire(import.meta.url);
            _native = require(join(searchPath, 'qks.node'));
            return _native;
        } catch (e) {
            lastError = e as Error;
            continue;
        }
    }

    throw new Error(
        `QKS native library not found. Searched: ${searchPaths.join(', ')}\n` +
        `Last error: ${lastError?.message}\n` +
        'Please build the Rust library with: cargo build --release'
    );
}

// Initialize native module
_native = loadNative();

// ============================================================================
// Type Definitions
// ============================================================================

/** Opaque handle to QKS instance */
export type QksHandle = Buffer;

/** QKS configuration */
export interface QksConfigNative {
    numQubits: number;
    useGpu: boolean;
    enableConsciousness: boolean;
    enableMetacognition: boolean;
}

/** IIT Φ computation result (native) */
export interface PhiResultNative {
    phi: number;
    mipSize: number;
    numStates: number;
    computationTimeMs: number;
}

/** Metacognitive introspection report (native) */
export interface IntrospectionReportNative {
    numBeliefs: number;
    numGoals: number;
    numCapabilities: number;
    confidence: number;
    timestampMs: bigint;
}

// ============================================================================
// Core Functions
// ============================================================================

/**
 * Create QKS plugin instance.
 */
export function qksCreate(config: QksConfigNative): QksHandle {
    return _native.qks_create(config);
}

/**
 * Destroy QKS plugin instance.
 */
export function qksDestroy(handle: QksHandle): void {
    _native.qks_destroy(handle);
}

/**
 * Get QKS version string.
 */
export function qksVersion(): string {
    return _native.qks_version();
}

// ============================================================================
// Layer 6: Consciousness
// ============================================================================

/**
 * Compute integrated information Φ (IIT 3.0).
 */
export function qksComputePhi(
    handle: QksHandle,
    networkState: Float64Array
): PhiResultNative {
    return _native.qks_compute_phi(handle, networkState);
}

/**
 * Broadcast content to Global Workspace.
 */
export function qksBroadcast(
    handle: QksHandle,
    content: Buffer,
    priority: number
): number {
    return _native.qks_broadcast(handle, content, priority);
}

// ============================================================================
// Layer 7: Metacognition
// ============================================================================

/**
 * Perform introspection of cognitive state.
 */
export function qksIntrospect(handle: QksHandle): IntrospectionReportNative {
    return _native.qks_introspect(handle);
}

/**
 * Monitor performance metrics.
 */
export function qksMonitorPerformance(
    handle: QksHandle,
    metrics: Float64Array
): void {
    _native.qks_monitor_performance(handle, metrics);
}

// ============================================================================
// Layer 8: Integration
// ============================================================================

/**
 * Execute one cognitive cycle.
 */
export function qksCognitiveCycle(
    handle: QksHandle,
    sensoryInput: Buffer
): Buffer {
    return _native.qks_cognitive_cycle(handle, sensoryInput);
}

// ============================================================================
// Error Handling
// ============================================================================

/**
 * Get last error message from native library.
 */
export function qksGetLastError(): string | null {
    try {
        return _native.qks_get_last_error();
    } catch {
        return null;
    }
}

/**
 * Check result code and throw error if failed.
 */
export function checkError(result: number): void {
    if (result !== 0) {
        const error = qksGetLastError();
        throw new Error(error || `QKS error code: ${result}`);
    }
}
